quickwit/quickwit-indexing/src/source/gcp_pubsub_source.rs

```diff
@@ -0,0 +1,355 @@
+
+        info!(
+            index_id=%ctx.index_uid.index_id(),
+            source_id=%ctx.source_config.source_id,
+            subscription=%subscription,
+            parallelism=%pull_parallelism,
+            "Starting GcpPubSub source."
+        );
+
+        Ok(GcpPubSubSource {
+            ctx,
+            subscription,
+            state: GcpPubSubSourceState::default(),
+            subscription_source,
+            backfill_mode_enabled,
+            publish_lock,
+            partition_id: Uuid::new_v4().to_string(),
+            pull_parallelism,
+            max_messages_per_pull,
+        })
+    }
+
+    fn should_exit(&self) -> bool {
+        self.backfill_mode_enabled && self.state.subscription_is_empty
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+struct BatchBuilder {
+    docs: Vec<Bytes>,
+    num_bytes: u64,
+    checkpoint_delta: SourceCheckpointDelta,
+}
+
+impl BatchBuilder {
+    fn build(self) -> RawDocBatch {
+        RawDocBatch {
+            docs: self.docs,
+            checkpoint_delta: self.checkpoint_delta,
+            force_commit: false,
+        }
+    }
+
+    fn build_force(self) -> RawDocBatch {
+        RawDocBatch {
+            docs: self.docs,
+            checkpoint_delta: self.checkpoint_delta,
+            force_commit: true,
+        }
+    }
+
+    // fn clear(&mut self) {
+    //     self.docs.clear();
+    //     self.num_bytes = 0;
+    //     self.checkpoint_delta = SourceCheckpointDelta::default();
+    // }
+
+    fn push(&mut self, message: ReceivedMessage) {
+        let doc = message.message.data;
+        self.num_bytes += doc.len() as u64;
+        self.docs.push(doc.into());
+    }
+}
+
+#[async_trait]
+impl Source for GcpPubSubSource {
+    async fn initialize(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<(), ActorExitStatus> {
+        info!("GcpPubSub initializing");
+        let publish_lock = self.publish_lock.clone();
+        ctx.send_message(doc_processor_mailbox, NewPublishLock(publish_lock))
+            .await?;
+        info!("GcpPubSub initialized");
+        Ok(())
+    }
+
+    async fn emit_batches(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<Duration, ActorExitStatus> {
+        info!("GcpPubSub beginning batch");
+        let now = Instant::now();
+        let batch_lock = Arc::new(RwLock::new(BatchBuilder::default()));
+        let deadline = time::sleep(*quickwit_actors::HEARTBEAT / 2);
+        tokio::pin!(deadline);
+
+        info!("GcpPubSub pulling batch");
+
+        loop {
+            let mut handles = vec![];
+
+            for _ in 1..self.pull_parallelism {
+                handles.push(self.pull_message_batch(Arc::clone(&batch_lock)));
+            }
+
+            tokio::select! {
+                _ = futures::future::join_all(handles) => {
```

Comment:
We should revisit how we handle pull parallelism in this source. Let's spawn one tokio task per `pull_parallelism` and have each task send a pull batch into a shared bounded channel. You can take a look at how it is done in the Kinesis source.